liquibase/liquibase
# Added supports method to DatabaseConnection

## Environment
**Liquibase Version**: 4.3.2

**Liquibase Integration & Version**: maven

**Liquibase Extension(s) & Version**: liquibase-mongodb-4.3.2

**Database Vendor & Version**: MongoDB 4.0.3

**Operating System Type & Version**: Windows

## Pull Request Type
* [ ] Enhancement/New feature (non-breaking change which adds functionality)

## Description
This is the initial fix in core. It is required a fix for the extensions as well. 
First commit introduces a support(Driver) method in DatabaseConnection interface with default implementation to return true. 
This is in order to not be a breaking change. Otherwise it is required to be implemented in any Implementations.

Subsequent commits has to fix Executor injection based on support(Database) implementation. This part has to be proved as not working after the initial commit.
Exception for the wrong Executor after Connection fixed:
`	at liquibase.nosql.executor.NoSqlExecutor.queryForObject(NoSqlExecutor.java:80)
	at liquibase.database.AbstractJdbcDatabase.getConnectionSchemaName(AbstractJdbcDatabase.java:364)
	at liquibase.database.AbstractJdbcDatabase.getDefaultSchemaName(AbstractJdbcDatabase.java:325)
	at liquibase.database.AbstractJdbcDatabase.escapeObjectName(AbstractJdbcDatabase.java:919)
	at liquibase.snapshot.SnapshotGeneratorFactory.has(SnapshotGeneratorFactory.java:114)
	at liquibase.snapshot.SnapshotGeneratorFactory.hasDatabaseChangeLogLockTable(SnapshotGeneratorFactory.java:279)`

Fixes #1782 
Fixes #1783 

## Steps To Reproduce
~~run update for a postgres or other db having liquibase~~mongodb.jar in classpath

## Actual Behavior
Exception described in the issues linked 

## Expected/Desired Behavior
Successful run

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: https://github.com/liquibase/liquibase/issues/1782 and https://github.com/liquibase/liquibase/issues/1783
* Local Branch: https://github.com/liquibase/liquibase/tree/LB-1412
* Unit Tests: https://github.com/liquibase/liquibase/pull/1784/checks
* Integration Tests: (shown in Unit Tests link above)
* Functional Tests: https://jenkins.datical.net/job/liquibase-pro/job/LB-1412/

#### Testing
* Steps to Reproduce: https://github.com/liquibase/liquibase/issues/1782 and https://github.com/liquibase/liquibase/issues/1783
* Guidance: 
  * This PR only impacts the liquibase API. Extensions need to use it in order to fix related issues

#### Dev Verification
Tested mongodb + postgresql using this PR plus https://github.com/liquibase/liquibase-mongodb/pull/148 which fixes the extension.
(No output saved from running it previously)

## Test Requirements (Internal Liquibase QA)
* No new automated functional test required for this change.

_Setup for Manual Test_

* Download the latest [liquibase-mongodb extension](https://github.com/liquibase/liquibase-mongodb) 
* Download the [MongoDB driver](https://mvnrepository.com/artifact/org.mongodb/mongo-java-driver)
* Put the liquibase-mongodb.jar into your liquibase/lib directory.
* Put the mongo-java-driver.jar into the liquibase/lib directory.
* Ensure you have a valid liquibase.properties file for a different (not Mongo) database platform, as well as the driver for the other (not Mongo) database platform.

_Verify Liquibase update does not use the MongoDB driver for a non-Mongo database connection._
_Verify Liquibase update is successful and has no errors related to a platform/driver mismatch._



┆Issue is synchronized with this [Jira Story](https://datical.atlassian.net/browse/LB-1412) by [Unito](https://www.unito.io)
